### ISSUE TITLE
redundant_clone: False positive with borrowed data

### ISSUE THREAD

**Issue by jgallagher-cs** [rust-lang/rust-clippy#5707]
```rust
use std::path::PathBuf;
use std::collections::HashMap;

fn main() {
    let p = PathBuf::from("/");

    let mut h: HashMap<&str, &str> = HashMap::new();
    h.insert("orig-p", p.to_str().unwrap());

    let mut q = p.clone();
    q.push("foo");

    println!("{:?} {}", h, q.display());
}
```

This warns about a redundant clone:

```
warning: redundant clone
  --> src/main.rs:10:18
   |
10 |     let mut q = p.clone();
   |                  ^^^^^^^^ help: remove this
   |
   = note: `#[warn(clippy::redundant_clone)]` on by default
note: this value is dropped without further use
  --> src/main.rs:10:17
   |
10 |     let mut q = p.clone();
   |                 ^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_clone
```

However, removing the clone as suggested is illegal, as we can't move `p` because of the reference stored in `h`:

```
error[E0505]: cannot move out of `p` because it is borrowed
  --> src/main.rs:10:17
   |
8  |     h.insert("orig-p", p.to_str().unwrap());
   |                        - borrow of `p` occurs here
9  |
10 |     let mut q = p;
   |                 ^ move out of `p` occurs here
...
13 |     println!("{:?} {}", h, q.display());
   |                         - borrow later used here

error: aborting due to previous error
```

cargo clippy -V: clippy 0.0.212 (d4092ace 2020-05-11)

**Comment by flip1995** [comment]
cc @sinkuu

**Comment by TjeuKayim** [comment]
I ran into a similar problem: when I apply clippy's suggestion to remove `.clone()`, I get `error[E0382]: borrow of moved value`. 
Checkout https://github.com/TjeuKayim/rusqlite/commit/a2458970cd5402bc6ac68ce3047f1c025754ab22 and run `cargo clippy --features bundled,deserialize --all-targets`
This probably has the same root cause as this issue.

**Comment by lengyijun** [comment]
The reason is: when meet a function:
```
_1=&mut y;
_2=& x;
_3=ff(_1,_2,...);
```
`redundant_clone`  only relate `_3` with `x`, it doesn't relate `x` with `y`.

However, in this issue,
```
hashmap.insert(...)
```
is desugared to
```
Hashmap::insert(&mut self, ...)
```

That's why `redundant_clone` doesn't find relation with hashmap and inserted value.
